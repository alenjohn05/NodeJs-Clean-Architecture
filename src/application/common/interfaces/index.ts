export * from './ILogger';
export * from './IPostsRepository';
