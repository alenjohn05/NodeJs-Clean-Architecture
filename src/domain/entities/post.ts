export class Post {
  public id?: number;
  public createdAt: Date;
  public published: boolean;
  public title: string;

  constructor(post: { id?: number; createdAt: Date; published: boolean; title: string }) {
    this.id = post.id;
    this.createdAt = post.createdAt;
    this.published = post.published;
    this.title = post.title;
  }
}
