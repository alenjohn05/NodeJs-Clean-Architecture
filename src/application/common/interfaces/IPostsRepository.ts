import { Post } from '@domain/entities';

export interface IPostsRepository {
  create(parameters: { post: Post }): Promise<{ id: number }>;
  delete(parameters: { id: number }): Promise<void>;
  getById(parameters: { id: number }): Promise<Post | null>;
  list(parameters: { pageNumber: number; pageSize: number }): Promise<{ count: number; posts: Array<Post> }>;
  update(parameters: { post: Post }): Promise<void>;
}
