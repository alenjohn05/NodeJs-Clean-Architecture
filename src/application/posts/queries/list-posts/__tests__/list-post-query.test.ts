import { mockDeep } from 'jest-mock-extended';
import { IPostsRepository } from '@application/common/interfaces';
import { ValidationException } from '@application/common/exceptions';
import { Post } from '@domain/entities';
import { makeListPostQuery } from '../list-posts-query';

describe('listPostsQuery', () => {
  function setup() {
    const postsRepository = mockDeep<IPostsRepository>();
    const listPostQuery = makeListPostQuery({ postsRepository });

    return {
      listPostQuery,
      postsRepository,
    };
  }

  test('lists posts', async () => {
    const { listPostQuery, postsRepository } = setup();

    postsRepository.list.mockResolvedValueOnce({
      count: 11,
      posts: [
        new Post({
          id: 11,
          createdAt: new Date(2022, 1, 1),
          published: false,
          title: 'Mock post',
        }),
      ],
    });

    const result = await listPostQuery({
      pageNumber: 2,
      pageSize: 10,
    });

    expect(result).toMatchSnapshot();
  });

  describe('given an invalid query', () => {
    test('validates and throws validation exception', async () => {
      const { listPostQuery } = setup();

      const result = listPostQuery({
        pageNumber: 1,
        pageSize: 1000, // Cannot exceed 50
      });

      await expect(result).rejects.toThrow(ValidationException);
    });
  });
});
