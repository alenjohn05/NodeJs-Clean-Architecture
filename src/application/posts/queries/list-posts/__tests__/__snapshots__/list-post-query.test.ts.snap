// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listPostsQuery lists posts 1`] = `
{
  "count": 11,
  "hasNextPage": false,
  "hasPreviousPage": true,
  "pageNumber": 2,
  "pageSize": 10,
  "posts": [
    {
      "id": 11,
      "published": false,
      "title": "Mock post",
    },
  ],
  "totalPages": 2,
}
`;
